### PR TITLE
fix AP algorithm

### PR DESCRIPTION
**Pull request by ekorman**

### CODE REVIEW

File: api/tests/functional-tests/backend/metrics/test_detection.py [hunk 1]
```diff
@@ -689,7 +617,7 @@ def _metric_to_dict(m) -> dict:
         },
         {
             "ious": iou_thresholds,
-            "value": 0.58,
+            "value": 0.592,
```
Comment (czaloom) [inline api/tests/functional-tests/backend/metrics/test_detection.py:620]: Why do these test values change?

File: api/tests/functional-tests/backend/metrics/test_detection.py [hunk 2]
```diff
@@ -711,7 +639,7 @@ def _metric_to_dict(m) -> dict:
     expected_mar_metrics = [
         {
             "ious": iou_thresholds,
-            "value": 0.652,
+            "value": 0.654,
```
Comment (czaloom) [inline api/tests/functional-tests/backend/metrics/test_detection.py:642]: Changed values?

File: api/valor_api/backend/metrics/detection.py [hunk 3]
```diff
@@ -583,6 +582,42 @@ def _annotation_type_to_column(
...
+        .filter(prediction_filter)
+        .predictions(as_subquery=False)
+        .where(models.Prediction.id.notin_(pd_set))  # type: ignore - SQLAlchemy type issue
+        .subquery()
+    ).all()
+
+    for pd_id, score, dataset_name, grouper_id in predictions:
+        if (
+            grouper_id in ranking and pd_id not in pd_set
+        ):  # how come grouper_id may not be in ranking??
```
Comment (czaloom) [inline api/valor_api/backend/metrics/detection.py:605]: nit: Are these comments still relevant?

File: api/tests/functional-tests/backend/metrics/test_detection.py (hunk 1 above)
Comment (ekorman) [reply api/tests/functional-tests/backend/metrics/test_detection.py:620]: gonnna dig into this

File: api/tests/functional-tests/backend/metrics/test_detection.py (hunk 2 above)
Comment (ekorman) [reply api/tests/functional-tests/backend/metrics/test_detection.py]: this is fixed now

File: integration_tests/client/metrics/test_detection.py [hunk 4]
```diff
@@ -1593,3 +1600,489 @@ def test_evaluate_detection_with_label_maps(
...
+        "label": {"key": "key", "value": "other value"},
+    }
+
+
+def test_evaluate_detection_false_negatives_two_images_one_only_with_different_class_high_confidence_of_fp(
+    db: Session, dataset_name: str, model_name: str, client: Client
+):
+    """In this test we have
+        1. An image with a matching groundtruth and prediction (same class, `"value"`, and high IOU)
+        2. A second image with a groundtruth annotation with clas `"other value"` and a prediction with higher confidence
```
[the remark this answers is not in the document]
Comment (ekorman) [reply integration_tests/client/metrics/test_detection.py:1992]: thanks, fixed

**Review by ntlind — APPROVED** [review]

**Review by czaloom — APPROVED** [review]

**Pull request merged**